# Repaso de Python

### Objetivo:
El objetivo de esta sesión es repasar los conceptos básicos de **Python** que son necesarios para desarrollar modelos de **Machine Learning** de manera efectiva.

## Contenidos:

0. **Python Básico**  
   Introducción a la sintaxis básica de Python, estructuras de control, funciones y manejo de excepciones.

1. **Objetos**  
   Conceptos fundamentales sobre programación orientada a objetos en Python: clases, instancias, herencia y encapsulamiento.

2. **Operaciones con `filter`, `map`, `reduce`**  
   Uso de estas funciones de orden superior para aplicar operaciones de transformación y filtrado de colecciones.

3. **Pandas**  
   Introducción a las estructuras de datos de `Pandas`, como `DataFrames` y `Series`, junto con operaciones de manipulación de datos.

4. **Matplotlib y Seaborn**  
   Visualización de datos utilizando `Matplotlib` y `Seaborn`, creación de gráficos personalizados y análisis visual de datos.

5. **Aplicaciones**  
   Aplicación de los conceptos revisados en la sesión para resolver problemas simples y preparar los datos para su uso en modelos de **Machine Learning**.

# Python Philosophy..

https://www.python.org/dev/peps/pep-0020/#abstract


### <a href="https://colab.research.google.com/github/CienciaDatosUdea/002_EstudiantesAprendizajeEstadistico/blob/main/semestre2024-2/Sesiones/Sesion_01a_python_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


In [ ]:
import time
import this

Hermoso es mejor que feo.

Explícito es mejor que implícito.

Lo simple es mejor que lo complejo.

Complejo es mejor que complicado.

Plano es mejor que anidado.

Disperso es mejor que denso.

La legibilidad cuenta.

Los casos especiales no son lo suficientemente especiales como para romper las reglas.

Aunque la practicidad le gana a la pureza.

Los errores nunca deben pasar en silencio.

A menos que se silencie explícitamente.

Frente a la ambigüedad, rechace la tentación de adivinar.

Debe haber una, y preferiblemente solo una, forma obvia de hacerlo.

Aunque esa manera puede no ser obvia al principio a menos que seas holandés.

Ahora es mejor que nunca.

Aunque nunca suele ser mejor que *ahora mismo* ahora.

Si la implementación es difícil de explicar, es una mala idea.

Si la implementación es fácil de explicar, puede ser una buena idea.

Los espacios de nombres son una gran idea, ¡hagamos más de esos!

# Functions, Loops, Conditionals, and list comprehensions



In [ ]:
# Determinar los numeros pares entre 1, 20
my_list = []
for number in range(0, 21):
  if number % 2 == 0:
    my_list.append(number)
my_list[1:]

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]

In [ ]:
my_list = [number for number in range(0, 21)  if number %2 ==0]
my_list[1:]

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]

In [ ]:
type(my_list[0])

int

# Optimización del Cómputo Científico

Las operaciones anteriores, aunque útiles, no son suficientemente eficientes para el cómputo científico intensivo. Si se desea realizar cálculos con un mayor rendimiento y reducir significativamente los tiempos de ejecución, es recomendable utilizar librerías especializadas como **NumPy** y otras herramientas optimizadas para el procesamiento numérico.

Estas bibliotecas están diseñadas para aprovechar mejor los recursos del sistema y permiten realizar operaciones en grandes volúmenes de datos de manera más eficiente.


## Numpy

In [ ]:
import numpy as np
def pares_numpy():
  N = 20
  V = np.zeros(int(N/2)+1)
  i=0
  for number in range(0, 21):
    if number % 2 == 0:
      V[i]=number
      i=i+1
  return V

In [ ]:
%%timeit -n 10000
pares_numpy()

1.23 µs ± 67.4 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


# Herramientas mas avanzadas

## Numba:

Numba permite paralelizar el código Python para CPU y GPU, a menudo con solo aplicar un decorador a la función. Numba también funciona con NumPy, una biblioteca de Python que se usa para realizar operaciones matemáticas complejas. Numba es una alternativa más simple y rápida que otras bibliotecas como Cython o TensorFlow, ya que no requiere aprender una nueva sintaxis ni tener un compilador de C/C++ instalado



In [ ]:
from numba import jit
@jit
def pares():
  N = 20
  V = np.zeros(int(N/2)+1)
  i=0
  for number in range(0, 21):
    if number % 2 == 0:
      V[i]=number
      i=i+1
  return V

In [ ]:
%%timeit -n 10000
pares()

570 ns ± 287 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 2.5 MB/s eta 0:00:0000:0100:010m
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 390.9/390.9 kB 118.8 kB/s eta 0:00:00a 0:00:01
  Using cached opt_einsum-3.3.0-py3-none-any.whl (65 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 11.9 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 70.0/70.0 MB 1.0 MB/s eta 0:00:0000:0100:02m
  Using cached numpy-1.26.4-cp310-cp310-macosx_11_0_arm64.whl (14.0 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.23.5
    Uninstalling numpy-1.23.5:
      Successfully uninstalled numpy-1.23.5
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gensim 4.3.0 requires FuzzyTM>=0.4.0, which is not installed.
torchvision 0.16.2 requires torch==2.1.2, but you have torch 2.2.0.post100 which is incompatible.
numba 0.56.4 requires n

In [ ]:
import jax
import jax.numpy as jnp

x = jnp.arange(10)
print(x)

[0 1 2 3 4 5 6 7 8 9]


In [ ]:
%%timeit -n 100

N = 20
V = jnp.zeros(int(N/2)+1)

i = 0
for number in jnp.arange(0, 21):
  if number % 2 == 0:
    V = V.at[i].set(number) # Actualiza el vector con el número impar
    i=i+1

2.14 ms ± 128 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


# Funciones en python

In [ ]:
def times_tables1():
    """
    Params:
      --

    Return:
      -- lst: List

    """
    lst = []
    for i in range(10):
        for j in range (10):
            lst.append(i*j)
    return lst

In [ ]:
%%timeit -n 10000
times_tables1()

4.39 µs ± 1.25 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [ ]:
def times_tables2():
    return [j*i for i in range(10) for j in range(10)]

In [ ]:
%%timeit -n 10000
times_tables2()

3.32 µs ± 972 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [ ]:
times_tables1() == times_tables2()

True

# Lambda function


In [ ]:
f = lambda x: (x+2)

In [ ]:
f(2)

4

## Magic command
https://ipython.readthedocs.io/en/stable/interactive/magics.html

## Pep 8
https://peps.python.org/pep-0008/

# Built-in Functions
https://docs.python.org/3/library/functions.html

## Function map

Permite aplicar una función a cada elemento de un iterable, como una lista o un diccionario, sin usar un bucle explícito. Esto puede hacer el código más conciso y legible.
Devuelve un objeto map que es un iterador, lo que significa que se puede usar en otras partes del programa o convertirlo en otro tipo de iterable, como una lista o una tupla.
Se puede combinar con otras funciones integradas, como filter, reduce o zip, para crear operaciones más complejas sobre los iterables.
Se puede usar con funciones definidas por el usuario, funciones lambda o funciones integradas, lo que ofrece una gran flexibilidad y expresividad.
Se puede usar con múltiples iterables como argumentos, lo que permite aplicar una función a varios elementos al mismo tiempo.

In [ ]:
L = [i**2 for i in range(0, 10)]
L

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

In [ ]:
# Programacion funcional
b = map(lambda x: x**2, range(10))
list(b)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

In [ ]:
def funcion_map_sin_jit():
  b = map(lambda x: x**2, range(10))
  return list(b)

In [ ]:
%%timeit -n 1000000
funcion_map_sin_jit()

1.79 µs ± 21.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [ ]:
@jit
def funcion_map():
  b = map(lambda x: x**2, range(10))
  return list(b)

In [ ]:
%%timeit -n 1000000
funcion_map()

200 ns ± 63.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [ ]:
a = [10.00, 11.00, 12.34, 2.0 ]
b = [9.8, 11.10, 12.34, 2.01 ]

In [ ]:
var = map(min, a, b)
var

In [ ]:
list(var)

[9.8, 11.0, 12.34, 2.0]

In [ ]:
var

In [ ]:
list(var)

[]

In [ ]:
# Este resultado no muestra nada, ¿porqué?.
#La variable var ya fue evaluada a través de elementos funcionales
for item in var:
  print(item)

Ejemplo 1:

Determinar los primeros 100 numeros impares empleando la funcion map

In [ ]:
# Escriba su codigo acá

In [ ]:
# @title Solucion
def impar(x):
  if(x%2!=0):
    return x

q = map(lambda x: 2*x+1 if (2*x+1)<=100 else 0 , range(50))

Ejemplo 2:
Otro ejemplo con la operación map: Obtener el titulo y apellido de la cada unas de las entradas de la lista:

people = ['Dr. Simon Einstein', 'Dr. Pedro Euler  ', 'Dr. Juan Tesla', 'Dr. Daniel Maxwell']

Ejemplo:
Dr Eistein, Dr Euler, Dr Maxwell



In [ ]:
# Escriba su codigo acá

In [ ]:
# @title  Solution
people = ['Dr. Simon Einstein', 'Dr. Pedro Euler  ', 'Dr. Juan Tesla', 'Dr. Daniel Maxwell']
people[0].split()


def split_title_and_name(person):
    title = person.split()[0]
    lastname = person.split()[-1]
    return f'{title} {lastname}'

last_names = map(split_title_and_name, people)
list(last_names)

a = []
for p in people:
  a.append(split_title_and_name(p))
a

['Dr. Einstein', 'Dr. Euler', 'Dr. Tesla', 'Dr. Maxwell']

In [ ]:
#list(q)

# Filter function


Filtros avanzados en python a través de programacion funcional

Ref = https://docs.hektorprofe.net/python/funcionalidades-avanzadas/funcion-filter/,

https://docs.python.org/3/library/functions.html#filter

In [ ]:
def multiple(numero):    # Primero declaramos una función condicional
    if numero % 5 == 0:  # Comprobamos si un numero es múltiple de cinco
        return True      # Sólo devolvemos True si lo es

numeros = [2, 5, 10, 23, 50, 33, 5000]
a = filter(multiple, numeros)

In [ ]:
a

In [ ]:
list(a)

[5, 10, 50, 5000]

In [ ]:
list(a)

[]

Diferencia con map

In [ ]:
a = map(multiple, numeros)
list(a)

[None, True, True, None, True, None, True]

# Funcion reduce

In [ ]:
from functools import reduce # Importa el módulo functools

a = [1, 2, 3, 4, 5]
b=reduce(lambda x,y:x+y, a)
b

15

In [ ]:
a = [1, 2, 3, 4, 5]
b=map(lambda x,y:x+y, a,a)
list(b)

[2, 4, 6, 8, 10]

# Objetos

In [ ]:
class auto:
  """
  Esta clase asigna un color y un tipo
  a un clase tipo carro
  """
  var = "taller de carros"

  def set_name_tipo(self, new_tipo):
    self.tipo = new_tipo

  def set_name_color(self, new_color ):
    self.color = new_color



In [ ]:
carro = auto()
carro.set_name_color="rojo"
carro.set_name_tipo="bus "
print(f"El carro es {carro.set_name_color} y es un  {carro.set_name_tipo} " )

El carro es rojo y es un  bus  


In [ ]:
class circulo(object):
  def __init__(self, R, posx, posy ):
    self.R1 = R
    self.posx = posx
    self.posy= posy

  def Area(self):
    A = np.pi*(self.R1)**2
    return A

  def perimetro(self):
    return 2*np.pi*self.R1


class circulo_(object):
  def __init__(self):
    self.R1 = None
    self.posx = None
    self.posy= None

  def Area(self):
    A = np.pi*(self.R1)**2
    return A

  def perimetro(self):
    return 2*np.pi*self.R1



In [ ]:
c = circulo(1, 0, 0)

In [ ]:
c.Area()
c.perimetro()

6.283185307179586

In [ ]:
cc=circulo_()

In [ ]:
cc.posx=1
cc.posy=1
cc.R1=1

In [ ]:
cc.R1

1

In [ ]:
cc.perimetro()

6.283185307179586

## funciones avanzadas con clases
Given a sentence,you task is build a iterator the words
Ref = https://www.youtube.com/watch?v=C3Z9lJXI6Qw&ab_channel=CoreySchafer

In [ ]:
class Sentence:
    def __init__(self, sentence):
        self.sentence = sentence
        self.index = 0
        self.words = self.sentence.split()

    def __iter__(self):
        return self

    def __next__(self):
        if self.index >= len(self.words):
            raise StopIteration
        index = self.index
        self.index += 1
        return self.words[index]

my_sentence = Sentence('This is a test')

print(next(my_sentence))
print(next(my_sentence))
print(next(my_sentence))
print(next(my_sentence))
#print(next(my_sentence))


This
is
a
test


In [ ]:
a = Sentence("hola mundo esta es una prueba")

In [ ]:
a.__next__()

'hola'

# Diccionarios

Elementos basicos con diccionarios

In [ ]:
students_class = { "Bob": "Physics","Alice": "Physics","Ana": "Biology" }

In [ ]:
for i, s in enumerate(students_class):
  print(i, s, students_class[s])


0 Bob Physics
1 Alice Physics
2 Ana Biology


In [ ]:
students_class.items()

dict_items([('Bob', 'Physics'), ('Alice', 'Physics'), ('Ana', 'Biology')])

Otra forma de iteraciones para los diccionarios a través del metodo items()

In [ ]:
for key, val in students_class.items():
  print(key, val)

Bob Physics
Alice Physics
Ana Biology


Accediendo a los valores del diccionario
Accediendo a las claves
- metodo keys()
- metodo values()

In [ ]:
print(students_class.values())
print(students_class.keys())

dict_values(['Physics', 'Physics', 'Biology'])
dict_keys(['Bob', 'Alice', 'Ana'])


# Pandas.

## Series
## Data Frame

In [ ]:
import pandas as pd
students_class = { "Bob": "Physics",
                  "Alice": "Chemistry",
                  "Ana": "Biology" }

In [ ]:
# Ndarray unidimensional con ejes etiquetados
s = pd.Series(students_class)
s

Bob        Physics
Alice    Chemistry
Ana        Biology
dtype: object

In [ ]:
# https://pandas.pydata.org/docs/reference/series.html
print(type(s.index))
s.index

<class 'pandas.core.indexes.base.Index'>


Index(['Bob', 'Alice', 'Ana'], dtype='object')

In [ ]:
#Forma de acceder a los elementos con el número del indice
s.iloc[2]

'Biology'

In [ ]:
#Forma de acceder a los indices
s.loc["Alice"]

'Chemistry'

In [ ]:
s.Bob

'Physics'

### Definición clave valor con enteros como clave.

In [ ]:
class_code = {99:"Physics",
              100:"Chemistry",
              101:"English" }

In [ ]:
s = pd.Series(class_code)

In [ ]:
s

99       Physics
100    Chemistry
101      English
dtype: object

In [ ]:
s.iloc[2]

'English'

In [ ]:
s.loc[99]

'Physics'

Tambien podemos definir el objeto Serie  a partir de una lista

In [ ]:
grades = pd.Series([8,7,10,1])

In [ ]:
grades

0     8
1     7
2    10
3     1
dtype: int64

In [ ]:
for i, g in enumerate(grades):
  print(i,g)

0 8
1 7
2 10
3 1


In [ ]:
grades.mean()

6.5

In [ ]:
grades.describe()

count     4.000000
mean      6.500000
std       3.872983
min       1.000000
25%       5.500000
50%       7.500000
75%       8.500000
max      10.000000
dtype: float64

Definicion a través de un  numpy array



In [ ]:
x = np.random.randint(0,20, 100)
random_s = pd.Series(x)

In [ ]:
random_s

0      0
1      6
2      0
3      6
4     15
      ..
95    10
96     4
97     7
98    16
99     1
Length: 100, dtype: int64

In [ ]:
random_s.head()

0     0
1     6
2     0
3     6
4    15
dtype: int64

In [ ]:
#Recorrido por las claves y valores, el metodo head es considerado para mostrar pocos valores

for index, values in random_s.head().iteritems():
  print(index, values)

AttributeError: 'Series' object has no attribute 'iteritems'

In [ ]:
%%timeit -n 100
x = np.random.randint(0,20, 100)
random_s = pd.Series(x)
random_s+=2 # OPeraciones vectoriales a todo el data frame, más eficiente.

#Comparar cuando se tiene un ciclo para realizar la suma, ¿cuál es mas eficiente?

49.4 µs ± 9.38 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Agregando nuevos valores con indices diferentes



In [ ]:
s = pd.Series([1,2,3,4,9])

In [ ]:
s.loc["nuevo"]=2

In [ ]:
s

0        1
1        2
2        3
3        4
4        9
nuevo    2
dtype: int64

In [ ]:
s.loc["nuevo"]

2

In [ ]:
s["nuevo"]

2

In [ ]:
s.iloc[-1]


2

Otra *forma* de definir una serie es a través de :

In [ ]:
juan_class = pd.Series(["a", "b","c"], index=["0","1","2"])

In [ ]:
juan_class

0    a
1    b
2    c
dtype: object

# Data Frame



Un DataFrame es una lista  de series



In [ ]:
d1 = { "Name":"Juan", "Topic":"Quantum Mechanics", "Score" : 10}
d2 = { "Name":"Pedro", "Topic":"statistical", "Score" : 10}
d3 = { "Name":"Ana", "Topic":"Clasical Mechanics", "Score" : 10}

record1 = pd.Series(d1)
record2 = pd.Series(d2)
record3 = pd.Series(d3)

In [ ]:
# indices con números enteros
df1 = pd.DataFrame( [record1, record2, record3] )
df1

Name               Topic  Score
0   Juan   Quantum Mechanics     10
1  Pedro         statistical     10
2    Ana  Clasical Mechanics     10

In [ ]:
# Asignando nombre a los indices
df2 = pd.DataFrame( [record1, record2, record3] , index = ["UdeA","Unal", "ITM"] )
df2

Name               Topic  Score
UdeA   Juan   Quantum Mechanics     10
Unal  Pedro         statistical     10
ITM     Ana  Clasical Mechanics     10

In [ ]:
#Accediendo a los indices por el nombre
df2.loc["UdeA"]

Name                  Juan
Topic    Quantum Mechanics
Score                   10
Name: UdeA, dtype: object

In [ ]:
#Accediendo a los indices por el numero
df2.iloc[0]

Name                  Juan
Topic    Quantum Mechanics
Score                   10
Name: UdeA, dtype: object

In [ ]:
#Accediendo a un elemento en particular
df2.loc["UdeA", "Name"]


'Juan'

In [ ]:
#Accediendo a algunas columnas del data frame
df2.loc[:, ["Name", "Topic"]]

Name               Topic
UdeA   Juan   Quantum Mechanics
Unal  Pedro         statistical
ITM     Ana  Clasical Mechanics

Se recomienda crear copias del data frame cuando se esta trabajando con pandas a traves del metodo copy() y no con el operador =, dado que se comparte el mismo espacio de memoria

In [ ]:
df2

Name               Topic  Score
UdeA   Juan   Quantum Mechanics     10
Unal  Pedro         statistical     10
ITM     Ana  Clasical Mechanics     10

In [ ]:
a = df2

In [ ]:
a

Name               Topic  Score
UdeA   Juan   Quantum Mechanics     10
Unal  Pedro         statistical     10
ITM     Ana  Clasical Mechanics     10

In [ ]:
a.loc["UdeA", "Name"] = "JuanB"

In [ ]:
a

Name               Topic  Score
UdeA  JuanB   Quantum Mechanics     10
Unal  Pedro         statistical     10
ITM     Ana  Clasical Mechanics     10

In [ ]:
df2

Name               Topic  Score
UdeA  JuanB   Quantum Mechanics     10
Unal  Pedro         statistical     10
ITM     Ana  Clasical Mechanics     10

In [ ]:
b = df2.copy()

Eliminacion de columnas

In [ ]:
del b["Topic"]

In [ ]:
b

Name  Score
UdeA  JuanB     10
Unal  Pedro     10
ITM     Ana     10

 Agregando nuevas columnas al data frame

In [ ]:
b["Nueva"] = [10, 8, 3]

In [ ]:
b

Name  Score  Nueva
UdeA  JuanB     10     10
Unal  Pedro     10      8
ITM     Ana     10      3

# Problemas de practica

Los siguientes problemas no son entregables en el curso.

1. https://github.com/ajcr/100-pandas-puzzles/blob/master/100-pandas-puzzles.ipynb


2.

Empleando  los siguientes tiempos:
```
t = np.linspace(0, 2, 1000)
```
2.1 Crear un data frame de pandas para la posicion $y=ho-0.5gt^2$, $g=9.8m/s$, $h = 100 m$
2.2 Adicione una nueva columna para la velocidad y la aceleración.

Construya un nuevo data frame desde el tiempo t=0.5s a 1.5s, solo con las posición como funcion del tiempo.



2.3 Crear una clase Fracción que represente un número racional con numerador y denominador. La clase debe tener un método constructor que reciba el numerador y el denominador como argumentos, y los almacene como atributos. La clase también debe tener métodos para simplificar la fracción, sumarla con otra fracción, restarla con otra fracción, multiplicarla por otra fracción, dividirla por otra fracción, compararla con otra fracción y convertirla en un número decimal. Puedes ver una posible solución en 1.

2.4 Crear una clase Punto que represente un punto en el plano cartesiano con coordenadas x e y. La clase debe tener un método constructor que reciba las coordenadas x e y como argumentos, y las almacene como atributos. La clase también debe tener métodos para calcular la distancia entre dos puntos, el ángulo entre dos puntos, el punto medio entre dos puntos y el punto simétrico respecto a otro punto. Puedes ver una posible solución en 2.


2.6 Escribe una función que tome una lista de palabras y devuelva una lista con la longitud de cada palabra. Usa la función map para aplicar tu función a la lista de entrada. Por ejemplo, si la lista de entrada es [“hola”, “mundo”, “python”], la lista de salida debe ser [4, 5, 6].


2.7 Escribe una función que tome una lista de números y devuelva una lista con el cuadrado de cada número. Usa la función map para aplicar tu función a la lista de entrada. Por ejemplo, si la lista de entrada es [1, 2, 3, 4, 5], la lista de salida debe ser [1, 4, 9, 16, 25].

# Anexos a la clase

## Jax en CPU, GPU y TPU

**Jax** es una biblioteca más rápida y potente que **NumPy** para el aprendizaje automático, especialmente cuando se trata de aprovechar el rendimiento de aceleradores como **GPU** y **TPU**. Sin embargo, es importante tener en cuenta que **Jax** presenta algunas limitaciones y desafíos, ya que **NumPy** es una biblioteca más madura y estable para cálculos numéricos generales.

### ¿Qué es Jax?

**Jax** es una biblioteca de Python que combina las capacidades de **Autograd** y **XLA** para facilitar la investigación en aprendizaje automático de alto rendimiento. Jax permite:

- Diferenciar automáticamente funciones nativas de Python y NumPy.
- Ejecutar funciones en **CPU**, **GPU** o **TPU**, optimizando el rendimiento según el hardware disponible.
- Aplicar transformaciones componibles como:
  - `grad`: Para calcular gradientes automáticos.
  - `vmap`: Para vectorizar funciones automáticamente.
  - `jit`: Para la compilación "just in time" que acelera la ejecución.
  - `pmap`: Para la paralelización en múltiples dispositivos.

### Componentes Clave

- **XLA (Accelerated Linear Algebra)**: Es un compilador que optimiza el código de Jax para ejecutarlo en aceleradores, como **GPU** y **TPU**, mejorando el rendimiento de operaciones intensivas.
  
- **Autograd**: Permite calcular gradientes automáticos, incluso en operaciones complejas como bucles, ramas, recursión, y cierres. Además, es capaz de calcular derivadas de orden superior (derivadas de derivadas).

### Instalación de Jax

Para usar Jax en diferentes dispositivos, puedes instalarlo con los siguientes comandos:

- Para **CPU**:

  ```bash
  !pip install "jax[cpu]"
    ```
- Para GPU con soporte CUDA:  
  ```bash
  !pip install "jax[cuda]" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html
  ```

- Para TPU:
  ```bash
   !pip install "jax[tpu]" -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
  ```
## Documentación

Para más información sobre **Jax**, puedes consultar la documentación oficial:

- [Documentación oficial de Jax](https://jax.readthedocs.io/en/latest/)
- [Jax 101: Guía de introducción](https://jax.readthedocs.io/en/latest/jax-101/index.html)


!pip install "jax[cpu]"